i6_models/parts/factored_hybrid/diphone.py

```diff
@@ -0,0 +1,167 @@
+
+
+@dataclass
+class DiphoneBackendV1Config(ModelConfiguration):
+    """
+    Attributes:
+        context_mix_mlp_dim: inner dimension of the context mixing MLP layers
+        context_mix_mlp_num_layers: how many hidden layers on the MLPs there should be
+        dropout: dropout probabilty
+        left_context_embedding_dim: embedding dimension of the left context
+            values. Good choice is in the order of n_contexts.
+        n_contexts: the number of raw phonemes/acoustic contexts
+        num_hmm_states_per_phone: the number of HMM states per phoneme
+        num_inputs: input dimension of the backend, must match w/ output dimension
+            of main encoder (e.g. Conformer)
+        phoneme_state_class: the phoneme state augmentation to apply
+        activation: activation function to use in the context mixing MLP.
+    """
+
+    left_context_embedding_dim: int
+    n_contexts: int
+    num_hmm_states_per_phone: int
+    phoneme_state_class: Union[int, PhonemeStateClassV1]
+
+    activation: Callable[[], nn.Module]
+    context_mix_mlp_dim: int
+    context_mix_mlp_num_layers: int
+    dropout: float
+    num_inputs: int
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+
+        assert self.context_mix_mlp_dim > 0
+        assert self.context_mix_mlp_num_layers > 0
+        assert self.n_contexts > 0
+        assert self.num_hmm_states_per_phone > 0
+        assert self.num_inputs > 0
+        assert self.left_context_embedding_dim > 0
+        assert 0.0 <= self.dropout <= 1.0, "dropout must be a probability"
+
+
+class DiphoneBackendV1(nn.Module):
+    """
+    Diphone FH model backend.
+
+    Consumes the output h(x) of a main encoder model and computes factored output
+    logits/probabilities for p(c|l,x) and p(l|x).
+    """
+
+    def __init__(self, cfg: DiphoneBackendV1Config):
+        super().__init__()
+
+        self.n_contexts = cfg.n_contexts
+
+        self.left_context_encoder = get_mlp(
+            num_input=cfg.num_inputs,
+            num_output=cfg.n_contexts,
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+        self.left_context_embedding = nn.Embedding(cfg.n_contexts, cfg.left_context_embedding_dim)
+        self.center_encoder = get_mlp(
+            num_input=cfg.num_inputs + cfg.left_context_embedding_dim,
+            num_output=get_center_dim(cfg.n_contexts, cfg.num_hmm_states_per_phone, cfg.phoneme_state_class),
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+
+    def forward(
+        self,
+        features: Tensor,  # B, T, F
+        contexts_left: Optional[Tensor] = None,  # B, T
+    ) -> Union[DiphoneLogitsV1, DiphoneProbsV1]:
```

Comment:
I know what you semantically want to say here, but since DiphoneProbsV1 inherits from DiphoneLogitsV1, programmatically I think this would be equivalent to
```suggestion
    ) -> DiphoneLogitsV1:
```

This change directly contradicts one other comment I made below, pls apply only one of those :P

Reply:
I just want to raise that "maybe" dataclasses as output might not be optimal for some graph tracing/scripting logic from PyTorch.

Reply:
Maybe a 
```
Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor], Optional[Tensor]]
```
with [joined, center, left, left_embedding] and depending on the conditions, each can be None if not applicable.?

Or should we just always return all of the above and then in training / inference some are just not used?

Reply:
I'm moving to tuples.